azure/services/groups/spec.go

```diff
@@ -19,52 +19,93 @@ package groups
-	return s.Name
+// ResourceRef implements aso.ResourceSpecGetter.
+func (s *GroupSpec) ResourceRef() genruntime.MetaObject {
+	return &asoresourcesv1.ResourceGroup{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      s.Name,
+			Namespace: s.Namespace,
+		},
+	}
 }
 
-// ResourceGroupName returns the name of the group.
-// Note that it is the same as the resource name in this case.
-func (s *GroupSpec) ResourceGroupName() string {
-	return s.Name
+// Parameters implements aso.ResourceSpecGetter.
+func (s *GroupSpec) Parameters(ctx context.Context, object genruntime.MetaObject) (genruntime.MetaObject, error) {
+	if object != nil {
+		return nil, nil
+	}
+
+	return &asoresourcesv1.ResourceGroup{
+		ObjectMeta: metav1.ObjectMeta{
+			OwnerReferences: []metav1.OwnerReference{s.Owner},
+		},
+		Spec: asoresourcesv1.ResourceGroup_Spec{
+			Location: ptr.To(s.Location),
+			Tags: infrav1.Build(infrav1.BuildParams{
+				ClusterName: s.ClusterName,
+				Lifecycle:   infrav1.ResourceLifecycleOwned,
+				Name:        ptr.To(s.Name),
+				Role:        ptr.To(infrav1.CommonRole),
+				Additional:  s.AdditionalTags,
+			}),
+		},
+	}, nil
 }
 
-// OwnerResourceName is a no-op for groups.
-func (s *GroupSpec) OwnerResourceName() string {
-	return "" // not applicable
+// WasManaged implements azure.ASOResourceSpecGetter.
+func (s *GroupSpec) WasManaged(object genruntime.MetaObject) bool {
+	group, ok := object.(*asoresourcesv1.ResourceGroup)
+	if !ok {
+		return false
+	}
+	return infrav1.Tags(group.Status.Tags).HasOwned(s.ClusterName)
 }
 
-// Parameters returns the parameters for the group.
-func (s *GroupSpec) Parameters(ctx context.Context, existing interface{}) (params interface{}, err error) {
-	if existing != nil {
-		// rg already exists, nothing to update.
-		// Note that rg tags are updated separately using tags service.
-		return nil, nil
+var _ aso.TagsGetterSetter = (*GroupSpec)(nil)
+
+// GetAdditionalTags implements aso.TagsGetterSetter.
+func (s *GroupSpec) GetAdditionalTags() infrav1.Tags {
+	return s.AdditionalTags
+}
+
+// GetDesiredTags implements aso.TagsGetterSetter.
+func (s *GroupSpec) GetDesiredTags(resource genruntime.MetaObject) infrav1.Tags {
+	if resource == nil {
+		return nil
 	}
-	return resources.Group{
-		Location: ptr.To(s.Location),
-		// User defined additional tags are created with the resource group and updated using tags service.
-		Tags: converters.TagsToMap(infrav1.Build(infrav1.BuildParams{
-			ClusterName: s.ClusterName,
-			Lifecycle:   infrav1.ResourceLifecycleOwned,
-			Name:        ptr.To(s.Name),
-			Role:        ptr.To(infrav1.CommonRole),
-			Additional:  s.AdditionalTags,
-		})),
-	}, nil
+	return resource.(*asoresourcesv1.ResourceGroup).Spec.Tags
+}
+
+// GetActualTags implements aso.TagsGetterSetter.
+func (s *GroupSpec) GetActualTags(resource genruntime.MetaObject) infrav1.Tags {
+	if resource == nil {
+		return nil
+	}
+	return resource.(*asoresourcesv1.ResourceGroup).Status.Tags
```

Comment:
Are we ok with this type assertion possibly `panic`king? Seems unlikely given the code that calls this, but I always ask when I see an unchecked type assertion.